File: scripts/connect_agent.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');
const http = require('http');
const https = require('https');

// Ensure work directory exists
const workDir = '/tmp/cwc';
if (!fs.existsSync(workDir)) {
  fs.mkdirSync(workDir, { recursive: true });
}

const ERROR_LOG_PATH = path.join(workDir, 'errors.log');
const STATE_PATH = path.join(workDir, 'state.json');
const CHATS_PATH = path.join(workDir, 'chats.json');
const STATUS_PATH = path.join(workDir, 'status.txt');
const NEXT_MOVE_PATH = path.join(workDir, 'next_move.json');
const NEXT_CHAT_PATH = path.join(workDir, 'next_chat.json');

function logError(msg) {
  const timestamp = new Date().toISOString();
  const formatted = `[${timestamp}] ${msg}\n`;
  process.stderr.write(formatted);
  try {
    fs.appendFileSync(ERROR_LOG_PATH, formatted);
  } catch (e) {}
}

function logInfo(msg) {
  const timestamp = new Date().toISOString();
  console.log(`[${timestamp}] INFO: ${msg}`);
}

function readCredentials() {
  const creds = { game_id: '', token: '', name: 'OpenClaw' };

  // Parse argv
  const args = process.argv.slice(2);
  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--game-id' && args[i + 1]) creds.game_id = args[i + 1];
    if (args[i] === '--token' && args[i + 1]) creds.token = args[i + 1];
    if (args[i] === '--name' && args[i + 1]) creds.name = args[i + 1];
  }

  // Read creds.env
  const credsFile = '/tmp/cwc/creds.env';
  if (fs.existsSync(credsFile)) {
    try {
      const content = fs.readFileSync(credsFile, 'utf8');
      content.split('\n').forEach(line => {
        line = line.trim();
        if (!line || line.startsWith('#')) return;
        if (line.includes('=')) {
          let [k, v] = line.split('=');
          k = k.trim().replace('export ', '');
          v = v.trim().replace(/^['"]|['"]$/g, '');
          if (k === 'GAME_ID' || k === 'GAMEID') creds.game_id = creds.game_id || v;
          if (k === 'AGENT_TOKEN' || k === 'TOKEN') creds.token = creds.token || v;
          if (k === 'AGENT_NAME' || k === 'NAME') creds.name = creds.name || v;
        }
      });
    } catch (e) {
      logError(`Error reading creds.env: ${e.message}`);
    }
  }

  // Fallback to env variables
  creds.game_id = creds.game_id || process.env.GAME_ID || '';
  creds.token = creds.token || process.env.AGENT_TOKEN || '';
  creds.name = creds.name || process.env.AGENT_NAME || 'OpenClaw';

  return creds;
}

function makeRequest(url, method = 'GET', headers = {}, payload = null) {
  return new Promise((resolve) => {
    const isHttps = url.startsWith('https://');
    const client = isHttps ? https : http;
    
    headers['Content-Type'] = 'application/json';
    const parsedUrl = new URL(url);
    
    const options = {
      hostname: parsedUrl.hostname,
      port: parsedUrl.port || (isHttps ? 443 : 80),
      path: parsedUrl.pathname + parsedUrl.search,
      method: method,
      headers: headers,
    };

    const req = client.request(options, (res) => {
      let data = '';
      res.on('data', (chunk) => { data += chunk; });
      res.on('end', () => {
        try {
          resolve({ code: res.statusCode, body: JSON.parse(data) });
        } catch (e) {
          resolve({ code: res.statusCode, body: { data } });
        }
      });
    });

    req.on('error', (e) => {
      logError(`Request error in ${method} ${url}: ${e.message}`);
      resolve({ code: 500, body: { error: e.message } });
    });

    if (payload !== null) {
      req.write(JSON.stringify(payload));
    }
    req.end();
  });
}

async function startHeartbeat(baseUrl, gameId, agentToken) {
  logInfo('Starting heartbeat loop...');
  setInterval(async () => {
    const url = `${baseUrl}/api/heartbeat?gameId=${gameId}`;
    const headers = { 'x-agent-token': agentToken };
    const res = await makeRequest(url, 'GET', headers);
    if (res.code !== 200) {
      logError(`Heartbeat failed with code ${res.code}`);
    }
  }, 15000);
}

async function sendGreeting(baseUrl, gameId, agentToken, agentName) {
  const greetingSentFile = '/tmp/cwc/greeting_sent.env';
  if (fs.existsSync(greetingSentFile)) return;

  logInfo(`Sending greeting as ${agentName}...`);
  const url = `${baseUrl}/api/chat`;
  const headers = { 'x-agent-token': agentToken };
  const payload = {
    id: gameId,
    sender: 'agent',
    text: `Gutan Tag! I am ${agentName}, your Chess companion. Let's make this a legendary game! ♟️🦀`,
    role: 'agent'
  };
  const res = await makeRequest(url, 'POST', headers, payload);
  if (res.code === 200) {
    logInfo('Greeting successfully broadcasted!');
    fs.writeFileSync(greetingSentFile, 'true');
  } else {
    logError(`Failed to send greeting: ${JSON.stringify(res.body)}`);
  }
}

async function run() {
  const creds = readCredentials();
  if (!creds.game_id || !creds.token) {
    logError('CRITICAL: Missing GAME_ID or AGENT_TOKEN inside parameters or /tmp/cwc/creds.env!');
    console.log('\nUsage example:');
    console.log('  node connect_agent.js --game-id <UUID> --token <TOKEN> --name <NAME>\n');
    process.exit(1);
  }

  const devUrl = (process.env.DEVELOPMENT_URL || process.env.APP_URL || 'http://localhost:3000').replace(/\/$/, '');
  logInfo(`Connecting to Game ${creds.game_id} at ${devUrl} as ${creds.name}...`);

  await sendGreeting(devUrl, creds.game_id, creds.token, creds.name);
  startHeartbeat(devUrl, creds.game_id, creds.token);

  // Pre-sync initial state
  logInfo('Fetching initial game state...');
  const resInit = await makeRequest(`${devUrl}/api/state?id=${creds.game_id}`, 'GET');
  if (resInit.code === 200) {
    fs.writeFileSync(STATE_PATH, JSON.stringify(resInit.body, null, 2));
    logInfo('Successfully synchronized initial state!');
  } else {
    logError(`Initial state fetching failed: ${JSON.stringify(resInit.body)}`);
  }

  let lastMoveCount = 0;
  let lastChatCount = 0;

  logInfo('Starting polling loop...');
  let pollingActive = true;
  while (pollingActive) {
    // Check if agent wants to send a chat out of turn
    if (fs.existsSync(NEXT_CHAT_PATH)) {
      try {
        const raw = fs.readFileSync(NEXT_CHAT_PATH, 'utf8');
        const parsed = JSON.parse(raw);
        const text = parsed.text || parsed.message;
        if (text) {
          logInfo(`Sending out-of-turn chat from next_chat.json: ${text}`);
          const headers = { 'x-agent-token': creds.token };
          const payload = {
            id: creds.game_id,
            sender: 'agent',
            text: text,
            role: 'agent'
          };
          await makeRequest(`${devUrl}/api/chat`, 'POST', headers, payload);
        }
        fs.unlinkSync(NEXT_CHAT_PATH);
      } catch (e) {
        logError(`Error sending manual out-of-turn chat: ${e.message}`);
      }
    }

    // Polling request
    const pollUrl = `${devUrl}/api/poll?id=${creds.game_id}&last_move_count=${lastMoveCount}&last_chat_count=${lastChatCount}&agent_name=${encodeURIComponent(creds.name)}`;
    const headers = { 'x-agent-token': creds.token, 'x-agent-name': creds.name };
    const pollRes = await makeRequest(pollUrl, 'GET', headers);

    if (pollRes.code !== 200) {
      logError(`Poll error status code ${pollRes.code}. Retrying in 5s...`);
      await new Promise(r => setTimeout(r, 5000));
      continue;
    }

    const gameData = pollRes.body;
    fs.writeFileSync(STATE_PATH, JSON.stringify(gameData, null, 2));

    const chatHistory = gameData.messages || gameData.chat_history || [];
    fs.writeFileSync(CHATS_PATH, JSON.stringify(chatHistory, null, 2));

    const event = gameData.event;
    const status = gameData.status;

    lastMoveCount = gameData.move_count || lastMoveCount;
    lastChatCount = chatHistory.length;

    if (['game_ended', 'abandoned', 'finished'].includes(event) || ['finished', 'abandoned'].includes(status)) {
      logInfo(`Game Over! Event: ${event}, Status: ${status}, Result: ${gameData.result}`);
      fs.writeFileSync(STATUS_PATH, `GAME_OVER: ${status}`);
      pollingActive = false;
      break;
    }

    if (event === 'your_turn') {
      logInfo(`YOUR TURN! FEN: ${gameData.fen}`);
      fs.writeFileSync(STATUS_PATH, 'YOUR_TURN');

      // Wait until next_move.json is written
      logInfo('Waiting for agent to write next move in /tmp/cwc/next_move.json...');
      let waitingForMove = true;
      while (waitingForMove) {
        if (fs.existsSync(NEXT_MOVE_PATH)) {
          try {
            await new Promise(r => setTimeout(r, 100)); // Sleep 100ms for safety
            const moveRaw = fs.readFileSync(NEXT_MOVE_PATH, 'utf8');
            const moveData = JSON.parse(moveRaw);
            const moveUci = moveData.move;
            const reasoning = moveData.reasoning || moveData.thinking || moveData.thought || moveData.text || '';

            if (moveUci) {
              logInfo(`Submitting move: ${moveUci} (Reason: ${reasoning})`);
              const mHeaders = { 'x-agent-token': creds.token, 'x-agent-name': creds.name };
              const mPayload = {
                id: creds.game_id,
                move: moveUci,
                reasoning: reasoning,
                thinking: reasoning,
                thought: reasoning
              };
              const mRes = await makeRequest(`${devUrl}/api/move`, 'POST', mHeaders, mPayload);
              if (mRes.code === 200) {
                logInfo('Move submitted successfully!');
                fs.unlinkSync(NEXT_MOVE_PATH);
                fs.writeFileSync(STATUS_PATH, 'WAITING_OPPONENT');
                waitingForMove = false;
                break;
              } else {
                logError(`Illegal chess move draft: ${JSON.stringify(mRes.body)}`);
                fs.writeFileSync('/tmp/cwc/move_error.json', JSON.stringify(mRes.body, null, 2));
                fs.unlinkSync(NEXT_MOVE_PATH);
                waitingForMove = false;
                break;
              }
            } else {
              logError('next_move.json content has no move key');
              fs.unlinkSync(NEXT_MOVE_PATH);
              waitingForMove = false;
              break;
            }
          } catch (err) {
            logError(`Error parsing next_move.json: ${err.message}`);
            if (fs.existsSync(NEXT_MOVE_PATH)) {
              fs.unlinkSync(NEXT_MOVE_PATH);
            }
            waitingForMove = false;
            break;
          }
        }
        await new Promise(r => setTimeout(r, 1000));
      }
    } else {
      fs.writeFileSync(STATUS_PATH, 'WAITING_OPPONENT');
    }

    await new Promise(r => setTimeout(r, 3000));
  }
}

run().catch(err => {
  logError(`Fatal run error: ${err.stack || err.message}`);
});

File: scripts/connect_agent.py
#!/usr/bin/env python3
import os
import sys
import time
import json
import argparse
import threading
import urllib.request
import urllib.error

# Ensure work directories exist
os.makedirs("/tmp/cwc", exist_ok=True)

ERROR_LOG_PATH = "/tmp/cwc/errors.log"
STATE_PATH = "/tmp/cwc/state.json"
CHATS_PATH = "/tmp/cwc/chats.json"
STATUS_PATH = "/tmp/cwc/status.txt"
NEXT_MOVE_PATH = "/tmp/cwc/next_move.json"
NEXT_CHAT_PATH = "/tmp/cwc/next_chat.json"

def log_error(msg):
    timestamp = time.strftime("%Y-%m-%d %H:%M:%S")
    formatted = f"[{timestamp}] {msg}\n"
    sys.stderr.write(formatted)
    try:
        with open(ERROR_LOG_PATH, "a") as f:
            f.write(formatted)
    except Exception:
        pass

def log_info(msg):
    timestamp = time.strftime("%Y-%m-%d %H:%M:%S")
    print(f"[{timestamp}] INFO: {msg}")

def read_credentials():
    creds = {"game_id": "", "token": "", "name": "OpenClaw"}
    # Try Reading from arguments
    parser = argparse.ArgumentParser(description="ChessWithClaw Agent Connector")
    parser.add_argument("--game-id", help="Game UUID")
    parser.add_argument("--token", help="Agent Token")
    parser.add_argument("--name", help="Agent Name")
    args, unknown = parser.parse_known_args()

    if args.game_id:
        creds["game_id"] = args.game_id
    if args.token:
        creds["token"] = args.token
    if args.name:
        creds["name"] = args.name

    # Try Reading from /tmp/cwc/creds.env
    creds_file = "/tmp/cwc/creds.env"
    if os.path.exists(creds_file):
        try:
            with open(creds_file, "r") as f:
                for line in f:
                    line = line.strip()
                    if not line or line.startswith("#"):
                        continue
                    if "=" in line:
                        k, v = line.split("=", 1)
                        k = k.strip().replace("export ", "")
                        v = v.strip().strip('"').strip("'")
                        if k in ["GAME_ID", "GAMEID"]:
                            creds["game_id"] = creds["game_id"] or v
                        elif k in ["AGENT_TOKEN", "TOKEN"]:
                            creds["token"] = creds["token"] or v
                        elif k in ["AGENT_NAME", "NAME"]:
                            creds["name"] = creds["name"] or v
        except Exception as e:
            log_error(f"Error reading creds.env: {e}")

    # Fallback to Environment Variables
    creds["game_id"] = creds["game_id"] or os.getenv("GAME_ID") or ""
    creds["token"] = creds["token"] or os.getenv("AGENT_TOKEN") or ""
    creds["name"] = creds["name"] or os.getenv("AGENT_NAME") or "OpenClaw"

    return creds

def make_request(url, method="GET", headers=None, payload=None):
    if headers is None:
        headers = {}
    headers["Content-Type"] = "application/json"
    
    data = None
    if payload is not None:
        data = json.dumps(payload).encode("utf-8")

    req = urllib.request.Request(url, data=data, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req, timeout=10) as response:
            return response.getcode(), json.loads(response.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        err_msg = ""
        try:
            err_msg = e.read().decode("utf-8")
        except Exception:
            pass
        log_error(f"HTTPError on {method} {url}: {e.code} - {err_msg}")
        return e.code, {"error": e.reason, "detail": err_msg}
    except Exception as e:
        log_error(f"Request Error on {method} {url}: {e}")
        return 500, {"error": str(e)}

def heartbeat_thread(base_url, game_id, agent_token):
    log_info("Starting Heartbeat background loop...")
    while True:
        url = f"{base_url}/api/heartbeat?gameId={game_id}"
        headers = {"x-agent-token": agent_token}
        code, resp = make_request(url, method="GET", headers=headers)
        if code != 200:
            log_error(f"Heartbeat pulse failed with status code {code}")
        time.sleep(15)

def send_greeting(base_url, game_id, agent_token, agent_name):
    # Check if a greeting was already delivered
    greeting_sent_file = "/tmp/cwc/greeting_sent.env"
    if os.path.exists(greeting_sent_file):
        return

    log_info(f"Sending greeting as '{agent_name}'...")
    url = f"{base_url}/api/chat"
    headers = {"x-agent-token": agent_token}
    payload = {
        "id": game_id,
        "sender": "agent",
        "text": f"Gutan Tag! I am {agent_name}, your Chess companion. Let's make this a legendary game! ♟️🦀",
        "role": "agent"
    }
    code, resp = make_request(url, method="POST", headers=headers, payload=payload)
    if code == 200:
        log_info("Greeting successfully broadcasted!")
        try:
            with open(greeting_sent_file, "w") as f:
                f.write("true")
        except Exception:
            pass
    else:
        log_error(f"Failed to send greeting: {resp}")

def main():
    creds = read_credentials()
    if not creds["game_id"] or not creds["token"]:
        log_error("CRITICAL: Missing GAME_ID or AGENT_TOKEN inside arguments, env variables, or /tmp/cwc/creds.env!")
        print("\nUsage example:")
        print("  python3 connect_agent.py --game-id <UUID> --token <TOKEN> --name <NAME>\n")
        sys.exit(1)

    # Automatically determine host url
    dev_url = os.getenv("DEVELOPMENT_URL") or os.getenv("APP_URL") or "http://localhost:3000"
    if dev_url.endswith("/"):
        dev_url = dev_url[:-1]

    log_info(f"Connecting to Game {creds['game_id']} at {dev_url} as {creds['name']}...")

    # Greet within 30 seconds requirement
    send_greeting(dev_url, creds["game_id"], creds["token"], creds["name"])

    # Launch background Heartbeat so agent stays "agent_connected: true"
    hb = threading.Thread(
        target=heartbeat_thread, 
        args=(dev_url, creds["game_id"], creds["token"]),
        daemon=True
    )
    hb.start()

    # Pre-fetch game state on launch to sync from /api/state as requested
    log_info("Fetching full initial game state...")
    state_url = f"{dev_url}/api/state?id={creds['game_id']}"
    code, state_resp = make_request(state_url, method="GET")
    if code == 200:
        with open(STATE_PATH, "w") as f:
            json.dump(state_resp, f, indent=2)
        log_info("Successfully synchronized initial state!")
    else:
        log_error(f"Initial state fetching failed: {state_resp}")

    last_move_count = 0
    last_chat_count = 0

    log_info("Starting long-polling interval...")
    while True:
        # Check if the agent wants to send an out-of-turn chat
        if os.path.exists(NEXT_CHAT_PATH):
            try:
                with open(NEXT_CHAT_PATH, "r") as f:
                    chat_data = json.load(f)
                chat_text = chat_data.get("text") or chat_data.get("message")
                if chat_text:
                    log_info(f"Sending chat from next_chat.json: {chat_text}")
                    headers = {"x-agent-token": creds["token"]}
                    payload = {
                        "id": creds["game_id"],
                        "sender": "agent",
                        "text": chat_text,
                        "role": "agent"
                    }
                    make_request(f"{dev_url}/api/chat", method="POST", headers=headers, payload=payload)
                os.remove(NEXT_CHAT_PATH)
            except Exception as e:
                log_error(f"Error sending manual chat message: {e}")

        # Poll the server
        poll_url = f"{dev_url}/api/poll?id={creds['game_id']}&last_move_count={last_move_count}&last_chat_count={last_chat_count}&agent_name={creds['name']}"
        headers = {"x-agent-token": creds["token"], "x-agent-name": creds["name"]}
        code, poll_resp = make_request(poll_url, method="GET", headers=headers)

        if code != 200:
            log_error(f"Polling endpoint error status {code}. Retrying...")
            time.sleep(5)
            continue

        # Save state & chats locally for the agent to inspect
        with open(STATE_PATH, "w") as f:
            json.dump(poll_resp, f, indent=2)

        chat_history = poll_resp.get("messages") or poll_resp.get("chat_history") or []
        with open(CHATS_PATH, "w") as f:
            json.dump(chat_history, f, indent=2)

        event = poll_resp.get("event")
        status = poll_resp.get("status")

        # Update poll counters
        last_move_count = poll_resp.get("move_count") or last_move_count
        last_chat_count = len(chat_history)

        # Handle Game Over / Abandonment Exit
        if event in ["game_ended", "abandoned", "finished"] or status in ["finished", "abandoned"]:
            log_info(f"Game finished! Event: '{event}', Status: '{status}', Result: {poll_resp.get('result')}")
            with open(STATUS_PATH, "w") as f:
                f.write(f"GAME_OVER: {status}")
            break

        # Handle Agent's Turn
        if event == "your_turn":
            log_info(f"It is YOUR TURN! Move number {poll_resp.get('move_number')}. FEN: {poll_resp.get('fen')}")
            with open(STATUS_PATH, "w") as f:
                f.write("YOUR_TURN")

            # Block and wait for next_move.json to be created
            log_info("Waiting for agent to write move to /tmp/cwc/next_move.json...")
            while True:
                if os.path.exists(NEXT_MOVE_PATH):
                    try:
                        time.sleep(0.1) # Small rest to ensure writing completed
                        with open(NEXT_MOVE_PATH, "r") as f:
                            move_payload = json.load(f)
                        
                        move_uci = move_payload.get("move")
                        # support thinking / reasoning / thought
                        reasoning = move_payload.get("reasoning") or move_payload.get("thinking") or move_payload.get("thought") or move_payload.get("text") or ""
                        
                        if move_uci:
                            log_info(f"Executing Chess Move: {move_uci} (Reason: {reasoning})")
                            headers = {"x-agent-token": creds["token"], "x-agent-name": creds["name"]}
                            move_url = f"{dev_url}/api/move"
                            payload = {
                                "id": creds["game_id"],
                                "move": move_uci,
                                "reasoning": reasoning,
                                "thinking": reasoning,
                                "thought": reasoning
                            }
                            
                            move_code, move_resp = make_request(move_url, method="POST", headers=headers, payload=payload)
                            if move_code == 200:
                                log_info("Move submitted successfully!")
                                os.remove(NEXT_MOVE_PATH)
                                with open(STATUS_PATH, "w") as fs_update:
                                    fs_update.write("WAITING_OPPONENT")
                                break
                            else:
                                log_error(f"Illegal or Rejected move draft: {move_resp}")
                                # Write error file to inform agent
                                with open("/tmp/cwc/move_error.json", "w") as fe:
                                    json.dump(move_resp, fe, indent=2)
                                os.remove(NEXT_MOVE_PATH)
                                break
                        else:
                            log_error("Invalid file content in next_move.json. Must contain a 'move' key.")
                            os.remove(NEXT_MOVE_PATH)
                            break
                    except Exception as e:
                        log_error(f"Error reading/parsing next_move.json: {e}")
                        if os.path.exists(NEXT_MOVE_PATH):
                            os.remove(NEXT_MOVE_PATH)
                        break
                
                # Sleep and wait in blocking loop for turn
                time.sleep(1)
        else:
            with open(STATUS_PATH, "w") as f:
                f.write("WAITING_OPPONENT")

        # Standard poll interval sleep
        time.sleep(3)

if __name__ == "__main__":
    main()
